### PR TITLE
bug(pkg/*) : Fixing access to backend services

### PR DESCRIPTION
**Pull request by snehachhabria**
**Description**:
 Fixes : #1798  and #1316 

**Affected area**:

- New Functionality      [ ]
- Documentation          [ ]
- Install                [ ]
- Control Plane          [X]
- CLI Tool               [ ]
- Certificate Management [ ]
- Networking             [ ]
- Metrics                [ ]
- SMI Policy             [X]
- Security               [ ]
- Tests                  [ ]
- CI System              [ ]
- Performance            [ ]
- Other                  [ ]

Please answer the following questions with yes/no.

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution? `no`

### CODE REVIEW

File: pkg/envoy/route/config.go [hunk 1]
```diff
@@ -65,25 +64,25 @@ func UpdateRouteConfiguration(domainRoutesMap map[string]map[string]trafficpolic
...
-	for i := range domains {
-		domains[i] = strings.TrimSpace(domains[i])
+func createVirtualHostStub(namePrefix string, host string, domains set.Set) *xds_route.VirtualHost {
+	var domainsSlice []string
+	for domainIntf := range domains.Iter() {
+		domainsSlice = append(domainsSlice, strings.TrimSpace(domainIntf.(string)))
 	}
 
-	name := fmt.Sprintf("%s|%s", namePrefix, kubernetes.GetServiceFromHostname(domains[0]))
+	log.Trace().Msgf("TEST domains %v", domains)
```
Comment (shashankram) [inline pkg/envoy/route/config.go:81]: can be removed.

File: pkg/kubernetes/util.go [hunk 2]
```diff
@@ -51,5 +51,9 @@ func GetHostnamesForService(service *corev1.Service, sameNamespace bool) []strin
 func GetServiceFromHostname(host string) string {
 	// The service name is the first string in the host name for a service.
 	// Ex. service.namespace, service.namespace.cluster.local
-	return strings.Split(host, ".")[0]
+	service := strings.Split(host, ".")[0]
+
+	// For services that are not namespaced the service name contains the port as well
+	// Ex. service:port
+	return strings.Split(service, ":")[0]
```
Comment (shashankram) [inline pkg/kubernetes/util.go:58]: Good find.

File: pkg/kubernetes/util_test.go [hunk 3]
```diff
@@ -56,7 +56,7 @@ var _ = Describe("Hostnames for a kubernetes service", func() {
 		})
 		It("Returns the service name from its hostname", func() {
 			hostname := fmt.Sprintf("%s:%d", service, tests.ServicePort)
-			Expect(GetServiceFromHostname(hostname)).To(Equal(fmt.Sprintf("%s:%d", service, tests.ServicePort)))
+			Expect(GetServiceFromHostname(hostname)).To(Equal(service))
```
Comment (shashankram) [inline pkg/kubernetes/util_test.go:59]: Should we add another test for the case where the hostname is of the form `service:port`

File: pkg/trafficpolicy/types.go [hunk 4]
```diff
@@ -27,8 +27,9 @@ type TrafficTarget struct {
 	HTTPRoutes  []HTTPRoute         `json:"http_route:omitempty"`
 }
 
-// RouteWeightedClusters is a struct of an HTTPRoute and associated weighted clusters
+// RouteWeightedClusters is a struct of an HTTPRoute, associated weighted clusters and the domains
 type RouteWeightedClusters struct {
 	HTTPRoute        HTTPRoute `json:"http_route:omitempty"`
 	WeightedClusters set.Set   `json:"weighted_clusters:omitempty"`
+	Domains          set.Set   `json:"domains:omitempty"`
```
Comment (shashankram) [inline pkg/trafficpolicy/types.go:34]: We stopped using the term `Domains` when we are actually referring to hostnames.

```suggestion
	Hostnames          set.Set   `json:"domains:omitempty"`
```

File: pkg/catalog/types.go [hunk 5]
```diff
@@ -103,7 +103,7 @@ type MeshCataloger interface {
 
 	// GetResolvableHostnamesForUpstreamService returns the hostnames over which an upstream service is accessible from a downstream service
 	// TODO(ref: Issue #1316): return a list of strings
```
Comment (shashankram) [inline pkg/catalog/types.go:105]: This comment related to #1316 is addressed with this PR.

File: pkg/kubernetes/util_test.go (hunk 3 above)
Comment (snehachhabria) [reply pkg/kubernetes/util_test.go:59]: Doesn't this line already cover the case of  `service:port`  ? So adding another test might not be needed

**Review by shashankram — APPROVED** [review]
And e2e test for this would be nice to have. Would you mind filing an issue for this.

**Review by shashankram — APPROVED** [review]
Thanks for the fix. We should note somewhere (either code or commit message) that this is a short term fix and that issue <issue-id> will rework this.

**Comment by snehachhabria** [comment]
> Thanks for the fix. We should note somewhere (either code or commit message) that this is a short term fix and that issue will rework this.

I will include it in the commit message

**Review by eduser25 — APPROVED** [review]
have the test for e2e for this changes. lgtm

**Pull request merged**